src/sentry/notifications/notificationcontroller.py

```diff
@@ -169,224 +171,223 @@ def _get_layered_setting_options(
             MutableMapping[
                 NotificationSettingEnum,
                 MutableMapping[ExternalProviderEnum, NotificationSettingsOptionEnum],
             ],
-        ],
-    ]:
-        """
-        Returns a mapping of all enabled notification setting providers for the enabled options.
-        Note that this includes default settings for any notification types that are not set.
-        """
+        ] = defaultdict(
+            lambda: defaultdict(
+                lambda: defaultdict(
+                    lambda: NotificationSettingsOptionEnum.DEFAULT
+                )  # Use lambda to return the default enum value
+            )
+        )
+        for recipient, recipient_options_map in setting_options_map.items():
+            # check actor type against recipient type
+            if actor_type:
+                if actor_type == ActorType.USER and recipient_is_team(recipient):
+                    continue
+                if actor_type == ActorType.TEAM and recipient_is_user(recipient):
+                    continue
 
-        setting_options = self._get_layered_setting_options(**kwargs)
-        setting_providers = self._get_layered_setting_providers(**kwargs)
+            for type in types_to_search:
+                option_value = recipient_options_map[type]
+                if option_value == NotificationSettingsOptionEnum.NEVER:
+                    continue
 
-        setting_option_and_providers: MutableMapping[
-            Recipient,
-            MutableMapping[
-                Scope,
-                MutableMapping[
-                    NotificationSettingEnum,
-                    MutableMapping[ExternalProviderEnum, NotificationSettingsOptionEnum],
-                ],
-            ],
-        ] = defaultdict(lambda: defaultdict(lambda: defaultdict(dict)))
-        for recipient, setting_option in setting_options.items():
-            for scope, setting in setting_option.items():
-                for type, option_value in setting.items():
-                    if option_value == NotificationSettingsOptionEnum.NEVER:
+                provider_options_map = setting_providers_map[recipient][type]
+                for provider, provider_value in provider_options_map.items():
+                    if provider_value == NotificationSettingsOptionEnum.NEVER:
                         continue
+                    # use the option value here as it has more specific information
+                    result[recipient][type][provider] = option_value
 
-                    recipient_providers = setting_providers[recipient][scope][type]
-                    for provider, provider_value in recipient_providers.items():
-                        if provider_value == NotificationSettingsOptionEnum.NEVER:
-                            continue
+        return result
 
-                        setting_option_and_providers[recipient][scope][type][
-                            provider
-                        ] = provider_value
```

Comment:
@snigdhas I believe this is the source of the bug. We need to use `option_value` instead of `provider_value`. `provider_value` is always `always`/`never` while `option_value` is what has the special cases of committed/subscribed. And in tests we set it the wrong way which is why we didn't catch it.